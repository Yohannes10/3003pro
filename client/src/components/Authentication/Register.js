
import React, { useState } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import avatar from '../Authentication/assets/profile.png'
import toast, { Toaster } from 'react-hot-toast';
import { useFormik } from 'formik';
import convertToBase64 from '../Authentication/helper/convert';
import styles from '../styles/Username.module.css';
import zxcvbn from 'zxcvbn';


const Register = ({handleRegister}) => {
  const navigate = useNavigate();
  const [file, setFile] = useState();
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [confirmPassword, setConfirmPassword] = useState(""); 
  const [email, setEmail] = useState("");
  

  const handleSubmit = (e) => {
    e.preventDefault();

    if (password !== confirmPassword) {
      alert("Passwords do not match.");
      return;
    }
    handleRegister(username, password, email,confirmPassword );
    setUsername("");
    setPassword("");
    setEmail("");
    setConfirmPassword(""); // Clear confirm password field after submission


  };


  const formik = useFormik({
    initialValues: {
      name: '',
      email: '',
      company: '',
      empID: '',
      role: '',
      password: ''
    },

    onSubmit: async values => {
      // Assign the profile image file (if any) to the form values
      values = await Object.assign(values, { profile: file || '' });
    }
  });

  /** formik doensn't support file upload so we need to create this handler */
  const onUpload = async e => {
    // Convert the selected file to base64 format
    const base64 = await convertToBase64(e.target.files[0]);
    setFile(base64);
  };

  return (
    <div className="container mx-auto">
      <Toaster position="top-center" reverseOrder={false} />

      <div className="flex justify-center items-center h-screen">
        <div className={styles.glass} style={{ width: '45%', height:'100%' }}>
          <div className="title flex flex-col items-center">
            {/* Render the title and description */}
            <h4 className="text-5xl font-bold">Register</h4>
            <span className="py-4 text-xl w-2/3 text-center text-gray-500">
              Happy to join you!
            </span>
          </div>

          <form className="py-8" onSubmit={handleSubmit}>
            <div className="profile flex justify-center py-1">
              {/* Render the profile image */}
              <label htmlFor="profile">
                <img
                  src={file || avatar}
                  className={`${styles.profile_img}`}
                  alt="avatar"
                />
              </label>
              <input onChange={onUpload} type="file" id="profile" name="profile" />
            </div>

            <div className="textbox flex flex-col items-center gap-2">
              {/* Render the form inputs */}
              <input
                className={styles.textbox}
                type="text"
                placeholder="Email*"
                onChange={(e) => setEmail(e.target.value)}
              />
              <input
                className={styles.textbox}
                type="text"
                placeholder="username*"
                onChange={(e) => setUsername(e.target.value)}
              />
              <input
                className={styles.textbox}
                type="password"
                placeholder="password*"
                onChange={(e) => setPassword(e.target.value)}
              />
                <input
          className={styles.textbox}
          type="password"
          placeholder="Confirm Password*"
          value={confirmPassword}
          onChange={(e) => setConfirmPassword(e.target.value)}
        />
              <select
                className={styles.textbox}
                onChange={formik.handleChange}
                value={formik.values.position}
                name="position"
                required
              >
                <option value="">Select Privileges*</option>
                <option value="Read">Read</option>
                <option value="Read and Write">Read and Write</option>
                <option value="Full Control">Full Control</option>
              </select>
              <input
                className={styles.textbox}
                type="text"
                placeholder="Company*"
              />
              <input
                className={styles.textbox}
                type="text"
                placeholder="EMPID"
              />
              <button className={styles.btn} type="submit">
                Register
              </button>
            </div>

            <div className="text-center py-4">
              {/* Render the login link */}
              <span className="text-gray-500">
                Already Register?{' '}
                <Link className="text-red-500" to="/">
                  Login Now
                </Link>
              </span>
            </div>
          </form>
        </div>
      </div>
    </div>
  );
}

export default Register;
