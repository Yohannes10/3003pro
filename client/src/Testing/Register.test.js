import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Register from '../components/Authentication/Register';

test('handles form submission', () => {
  const mockHandleRegister = jest.fn();

  const { asFragment } = render(
    <MemoryRouter>
      <Register handleRegister={mockHandleRegister} />
    </MemoryRouter>
  );

  const emailInput = screen.getByPlaceholderText('Email*');
  const usernameInput = screen.getByPlaceholderText('username*');
  const passwordInput = screen.getByPlaceholderText('password*');
  const confirmPasswordInput = screen.getByPlaceholderText('Confirm Password*');

  fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
  fireEvent.change(usernameInput, { target: { value: 'testuser' } });
  fireEvent.change(passwordInput, { target: { value: 'testpassword' } });
  fireEvent.change(confirmPasswordInput, { target: { value: 'testpassword' } });

  const registerButton = screen.getByRole('button', { name: /register/i });
  fireEvent.click(registerButton);

  expect(mockHandleRegister).toHaveBeenCalledWith(
    'testuser',
    'testpassword',
    'test@example.com',
    'testpassword'
  );

  // Perform snapshot testing
  expect(asFragment()).toMatchSnapshot();
});
