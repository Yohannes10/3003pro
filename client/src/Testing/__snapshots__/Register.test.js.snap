// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles form submission 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto"
  >
    <div
      style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
    />
    <div
      class="flex justify-center items-center h-screen"
    >
      <div
        class="glass"
        style="width: 45%; height: 100%;"
      >
        <div
          class="title flex flex-col items-center"
        >
          <h4
            class="text-5xl font-bold"
          >
            Register
          </h4>
          <span
            class="py-4 text-xl w-2/3 text-center text-gray-500"
          >
            Happy to join you!
          </span>
        </div>
        <form
          class="py-8"
        >
          <div
            class="profile flex justify-center py-1"
          >
            <label
              for="profile"
            >
              <img
                alt="avatar"
                class="profile_img"
                src="profile.png"
              />
            </label>
            <input
              id="profile"
              name="profile"
              type="file"
            />
          </div>
          <div
            class="textbox flex flex-col items-center gap-2"
          >
            <input
              class="textbox"
              placeholder="Email*"
              type="text"
            />
            <input
              class="textbox"
              placeholder="username*"
              type="text"
            />
            <input
              class="textbox"
              placeholder="password*"
              type="password"
            />
            <input
              class="textbox"
              placeholder="Confirm Password*"
              type="password"
              value=""
            />
            <select
              class="textbox"
              name="position"
              required=""
            >
              <option
                value=""
              >
                Select Privileges*
              </option>
              <option
                value="Read"
              >
                Read
              </option>
              <option
                value="Read and Write"
              >
                Read and Write
              </option>
              <option
                value="Full Control"
              >
                Full Control
              </option>
            </select>
            <input
              class="textbox"
              placeholder="Company*"
              type="text"
            />
            <input
              class="textbox"
              placeholder="EMPID"
              type="text"
            />
            <button
              class="btn"
              type="submit"
            >
              Register
            </button>
          </div>
          <div
            class="text-center py-4"
          >
            <span
              class="text-gray-500"
            >
              Already Register? 
              <a
                class="text-red-500"
                href="/"
              >
                Login Now
              </a>
            </span>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders Login component 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto"
  >
    <div
      style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
    />
    <div
      class="flex justify-center items-center h-screen"
    >
      <div
        class="glass"
        style="width: 45%; height: 90%;"
      >
        <div
          class="title flex flex-col items-center"
        >
          <h4
            class="text-5xl font-bold"
          >
            Hello
          </h4>
          <span
            class="py-4 text-xl w-2/3 text-center text-gray-500"
          >
            Explore More by connecting with us.
          </span>
        </div>
        <form
          class="py-1"
        >
          <div
            class="profile flex justify-center py-4"
          >
            <img
              alt="avatar"
              class="profile_img"
              src="profile.png"
            />
          </div>
          <div
            class="textbox flex flex-col items-center gap-6"
          >
            <input
              class="textbox"
              placeholder="Username"
              type="text"
            />
            <input
              class="textbox"
              placeholder="Password"
              type="text"
            />
            <button
              class="btn"
              type="submit"
            >
              Sign In
            </button>
          </div>
          <div
            class="text-center py-4"
          >
            <span
              class="text-gray-500"
            >
              Forgot Password? 
              <a
                class="text-red-500"
                href="/"
              >
                Recover Now
              </a>
            </span>
          </div>
          <div
            class="text-center py-4"
          >
            <span
              class="text-gray-500"
            >
              Not a Member 
              <a
                class="text-red-500"
                href="/register"
              >
                Register Now
              </a>
            </span>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
